File: frontend/src/types/bill.ts
// Mirrors backend/app/schemas/bill.py
// NOTE: numeric fields are typed `string` because FastAPI/Pydantic
// serializes Decimal as a JSON string (to avoid float precision loss) —
// they are NOT native JSON numbers. Always wrap with Number(v) before
// doing arithmetic on two of these fields together.

export type SourceFormat = "pdf" | "xls";

export interface BillPeriod {
  id: string;
  label: string;
  corporate_code: string | null;
  bill_period_start: string | null;
  bill_period_end: string | null;
  invoice_date: string | null;
  stated_total_charges_for_bill_period: string | null;
  stated_total_due_amount: string | null;
  source_format: SourceFormat;
  reconciled: boolean;
  reconciliation_discrepancy: string | null;
  bucket_cost_override: string | null;
  bucket_vat_override: string | null;
  created_at: string;
}

export interface ImportResult {
  bill_period_id: string;
  line_items_imported: number;
  parsed_total_charges_for_bill_period: string;
  stated_total_charges_for_bill_period: string | null;
  reconciled: boolean;
  reconciliation_discrepancy: string | null;
  source_format: SourceFormat;
}

export interface BillSummaryRow {
  bill_line_item_id: string;
  mobile_no: string;

  emp_no: string | null;
  name: string | null;
  lob: string | null;
  cadre: string | null;
  credit_limit: string | null;
  level: string | null;
  email: string | null;
  project_label: string | null;

  total_usage_charges: string;
  // Only populated when the bill was imported from the .xls source —
  // null for PDF-sourced imports, which only give the combined total above.
  voice_rental: string | null;
  voice_usage: string | null;
  sms: string | null;
  data_rental: string | null;
  data_usage: string | null;

  idd: string;
  roaming: string;
  vas: string;
  charges_for_bill_period: string;
  vat: string;
  add_to_bill_charges: string;

  net_amount: string;
  bucket_cost: string;
  bucket_vat: string;
  bucket_nett: string;
  total: string;
  salary_deduction: string;
  need_approval: string;
  is_overridden: boolean;
  is_general_line: boolean;
  is_bucket_excluded: boolean;
}

export interface ApprovalOverrideInput {
  approval_override: string | null;
}

export interface BucketExclusionInput {
  is_bucket_excluded: boolean;
}

export interface BucketRateOverrideInput {
  bucket_cost_override: number | null;
  bucket_vat_override: number | null;
}

File: backend/app/schemas/bill.py
import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel


class BillPeriodOut(BaseModel):
    id: uuid.UUID
    label: str
    corporate_code: str | None
    bill_period_start: date | None
    bill_period_end: date | None
    invoice_date: date | None
    stated_total_charges_for_bill_period: Decimal | None
    stated_total_due_amount: Decimal | None
    source_format: str
    reconciled: bool
    reconciliation_discrepancy: Decimal | None
    bucket_cost_override: Decimal | None
    bucket_vat_override: Decimal | None
    created_at: datetime

    model_config = {"from_attributes": True}


class BucketRateOverrideInput(BaseModel):
    # Both null clears the override, reverting this bill period back to
    # whatever rate is active in the standard rate table.
    bucket_cost_override: Decimal | None = None
    bucket_vat_override: Decimal | None = None


class ImportResult(BaseModel):
    bill_period_id: uuid.UUID
    line_items_imported: int
    parsed_total_charges_for_bill_period: Decimal
    stated_total_charges_for_bill_period: Decimal | None
    reconciled: bool
    reconciliation_discrepancy: Decimal | None
    source_format: str


class ApprovalOverrideInput(BaseModel):
    approval_override: str | None  # e.g. "Manager approved", or null to clear it


class BucketExclusionInput(BaseModel):
    is_bucket_excluded: bool


class BillSummaryRow(BaseModel):
    """
    Mirrors the source Excel 'Summary' tab, as closely as the source file
    allows. Everything here is either taken straight from the bill or
    computed from a documented rule — nothing is guessed.
    """

    bill_line_item_id: uuid.UUID
    mobile_no: str

    emp_no: str | None
    name: str | None
    lob: str | None
    cadre: str | None
    credit_limit: Decimal | None
    level: str | None
    email: str | None
    project_label: str | None

    total_usage_charges: Decimal
    # Only populated when the bill was imported from the .xls source —
    # None for PDF-sourced imports, which only give the combined total above.
    voice_rental: Decimal | None
    voice_usage: Decimal | None
    sms: Decimal | None
    data_rental: Decimal | None
    data_usage: Decimal | None

    idd: Decimal
    roaming: Decimal
    vas: Decimal
    charges_for_bill_period: Decimal
    vat: Decimal
    add_to_bill_charges: Decimal

    net_amount: Decimal
    bucket_cost: Decimal
    bucket_vat: Decimal
    bucket_nett: Decimal
    total: Decimal
    salary_deduction: Decimal
    need_approval: str
    is_overridden: bool
    is_general_line: bool
    is_bucket_excluded: bool